### PR TITLE
Add XCBConnection::Handler

### PR DESCRIPTION
**Pull request by wmww**
Stacked on #1728. This simplifies passing around `on_success`/`on_error` callbacks. Starting to look similar to a future but I think it's still fine to have an XCB-specific concept here.

### CODE REVIEW

File: src/server/frontend_xwayland/xcb_connection.h [hunk 1]
```diff
@@ -90,6 +91,46 @@ class XCBConnection
...
+            if (error)
+            {
+                free(error);
+            }
+        }
+
+        operator xcb_generic_error_t**() { return &error; }
+
+        xcb_generic_error_t* error{nullptr};
+    };
```
Comment (AlanGriffiths) [inline src/server/frontend_xwayland/xcb_connection.h:107]: Looks as though ErrorPtr has incorrect CopyAssign semantics. (Is this a case for `UniqueCPtr`?)

File: src/server/frontend_xwayland/xcb_connection.cpp [hunk 2]
```diff
@@ -231,18 +230,23 @@ auto mf::XCBConnection::read_property(
         0, // no offset
         2048); // big buffer
 
-    return [this, cookie, action, on_error, window, prop]()
+    return [this, cookie, handler=std::move(handler), window, prop]()
         {
             try
             {
+                ErrorPtr error;
```
Comment (AlanGriffiths) [inline src/server/frontend_xwayland/xcb_connection.cpp:237]: Looks like this is default initialized and not written before it's read. Can that be right?

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
There's something odd about `ErrorPtr` and the way it is used. Otherwise, this looks good.

File: src/server/frontend_xwayland/xcb_connection.h (hunk 1 above)
Comment (wmww) [reply src/server/frontend_xwayland/xcb_connection.h:107]: I don't think you can get mutable access to the underlying pointer of a `unique_ptr`, right? The error has to be initialized as null, and then optionally set by the XCB function.

File: src/server/frontend_xwayland/xcb_connection.cpp (hunk 2 above)
Comment (wmww) [reply src/server/frontend_xwayland/xcb_connection.cpp]: Oops, actual usage got dropped in a rebase somewhere along the line.

**Review by AlanGriffiths — APPROVED** [review]
OK now

bors r+

**Pull request merged**